source/mod_geometry.f90

```diff
@@ -0,0 +1,808 @@
+
+  use parpro
+  use mod_common, only : iu, ic
+
+  integer, intent(in) :: iEl
+  integer, intent(in) :: ixEl
+
+  integer i, ix
+
+  geom_checkSingElemUnique = -1
+
+  do i=1,iu
+    ix = ic(i)-nblo
+    if(ix > 0) then ! SINGLE ELEMENT
+      if(i /= iEl .and. ix == ixEl) then
+        geom_checkSingElemUnique = i
+        return
+      end if
+    end if
+  end do
+
+end function geom_checkSingElemUnique
+
+! ================================================================================================ !
+!  A.Mereghetti, V.K. Berglyd Olsen, BE-ABP-HSS
+!  Check that a given entry in the sequence is unique
+!  Updated: 2019-03-28
+!  sLoc     : S-coordinate where element should be
+!  isLast   : True if return last lens at sLoc
+!  iEl      : Index in lattice structure of found element
+!  ixEl     : Index in array of SINGLE ELEMENTs of found element
+!  wasFound : True if element was found
+! ================================================================================================ !
+subroutine geom_findElemAtLoc(sLoc, isLast, iEl, ixEl, wasFound)
+
+  use parpro
+  use crcoall
+  use floatPrecision
+  use mod_common, only : iu, tlen, ic, dcum
+  use numerical_constants, only : zero
+
+  real(kind=fPrec), intent(out) :: sLoc
+  logical,          intent(in)  :: isLast
+  integer,          intent(out) :: iEl
+  integer,          intent(out) :: ixEl
+  logical,          intent(out) :: wasFound
+
+  integer i, iDelta, iCheck, iMax, iStep
+  logical lSlide
+
+  iEl  = -1
+  ixEl = -1
+
+  if(sLoc > tlen .or. sLoc < zero) then
+    write(lout,"(a,2(f11.4,a))") "GEOMETRY> ERROR Find Element: "//&
+      "Requested s-location: ",sLoc," is not inside accelerator range [0:",tlen,"]"
+    call prror
+  endif
+
+  ! Fast search
+  iCheck = iu
+  iDelta = iu
+  do while(iDelta > 1 .or. iCheck > 0 .or. iCheck < iu)
+    if(dcum(iCheck) == sLoc) exit
+    iDelta = nint(real(iDelta/2))
+    if(dcum(iCheck) < sLoc) then
+      iCheck = iCheck + iDelta
+    else
+      iCheck = iCheck - iDelta
+    end if
+  end do
+
+  ! Finalise search
+  if(dcum(iCheck) < sLoc .or. dcum(iCheck) == sLoc .and. isLast) then
+    iMax   = iu
+    iStep  = 1
+    lslide = isLast
+  else
+    iMax   = 1
+    iStep  = -1
+    lSlide = .not.isLast
+  endif
+  do i=iCheck,iMax,iStep
+    if(dcum(i) < sLoc) cycle
+    if(dcum(i) > sLoc) then
+      if(iEl == -1) iEl = i
+    else
+      iEl = i
+      wasFound = .true.
+      if(lSlide) cycle
+    end if
+    exit
+  end do
+
+  if(wasFound) then
+    ixEl = ic(iEl)-nblo
+    if(ixEl < 0) ixEl=ic(iEl) ! drift
+  else
+    write(lout,"(a,2(f11.4,a))") "GEOMETRY> Find Element: s-location: ",sLoc," was not found in acclerator range [0:",tlen,"]"
+  end if
+
+end subroutine geom_findElemAtLoc
+
+! ================================================================================================ !
+!  A. Mereghetti, D. Sinuela Pastor, for the FLUKA Team
+!  V.K. Berglyd Olsen, BE-ABP-HSS
+!  Original: 2014-06-13
```

Comment:
A small modification (since this is a sort of famous routine), to facilitate grepping:
`!  Original: 2014-06-13 (then called 'cadcum')`